### PR TITLE
added celery-django-project

### PR DESCRIPTION
**Pull request by shimpidipali**
sample project of celery-django-project
  - how to run project details in redme.md file.
  - how celery works
  - how to create multiple queue.
  - how celery works with redis.
  - how flower and curses uses as monitoring tool.

### CODE REVIEW

File: celery_with_django/django_celery_project/settings.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+Generated by 'django-admin startproject' using Django 5.0.3.
+
+For more information on this file, see
+https://docs.djangoproject.com/en/5.0/topics/settings/
+
+For the full list of settings and their values, see
+https://docs.djangoproject.com/en/5.0/ref/settings/
+"""
+
+from pathlib import Path
```
Comment (reverseBitsTech) [inline celery_with_django/django_celery_project/settings.py:13]: Install the import optimizer plugin @shimpidipali I will help you out

File: celery_with_django/django_celery_project/settings.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/5.0/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'django-insecure-w63dld2gxku7hkplgs%2i4hj3f0t_t8vrxxx0x6$u=pcq=g#7u'
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
+
+ALLOWED_HOSTS = []
```
Comment (reverseBitsTech) [inline celery_with_django/django_celery_project/settings.py:33]: Will this allow localhost by default? Also when we push the code or deploy the Debug will be `False`, I think in that case you need to allow the hosts, For now this is fine. Just keep note

File: celery_with_django/main_app/tasks.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from celery import shared_task
+from time import sleep
+
+@shared_task(bind=True)
+def add(self,x,y):
```
Comment (reverseBitsTech) [inline celery_with_django/main_app/tasks.py]: Can you add `x:int, y:int` type definitions here?

File: celery_with_django/main_app/tasks.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    sleep(60)
+    return x + y
+
+@shared_task(bind=True)
+def sub(self,x,y):
```
Comment (reverseBitsTech) [inline celery_with_django/main_app/tasks.py]:
```
def sub(self, x:int, y:int) -> int:
    # For observing the status change.
    sleep(120)
    return x - y
```

File: celery_with_django/main_app/views.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+from django.http import HttpResponse
+from .tasks import add, sub
+from send_mail_app.tasks import send_mail_func
+
+def test(request):
+    result_add = add.delay(10,20)
```
Comment (reverseBitsTech) [inline celery_with_django/main_app/views.py]: Don't think you need delay here. As you put the task `sleep` call inside the task.

Also Can you document a comment how the `delay` works here?

File: celery_with_django/run_celery.bat [hunk 6]
```diff
@@ -0,0 +1,13 @@
+@echo off
+
+rem Activate virtual environment
+call D:\ReverseBits\Projects\django\celery_with_django\env\Scripts\activate.bat
```
Comment (reverseBitsTech) [inline celery_with_django/run_celery.bat:4]: this should be picked up from Env as this same folder structure won't be available to others, or it should be a relative path for more dynamic calling

File: celery_with_django/main_app/tasks.py (hunk 3 above)
Comment (shimpidipali) [reply celery_with_django/main_app/tasks.py]: yes, I added.

File: celery_with_django/main_app/tasks.py (hunk 4 above)
Comment (shimpidipali) [reply celery_with_django/main_app/tasks.py]: added response Type.

File: celery_with_django/django_celery_project/settings.py (hunk 1 above)
Comment (shimpidipali) [reply celery_with_django/django_celery_project/settings.py:13]: ok

File: celery_with_django/django_celery_project/settings.py (hunk 2 above)
Comment (shimpidipali) [reply celery_with_django/django_celery_project/settings.py:33]: ok, I keep this note

File: celery_with_django/main_app/views.py (hunk 5 above)
Comment (shimpidipali) [reply celery_with_django/main_app/views.py]: added comments how delay works.

File: celery_with_django/run_celery.bat (hunk 6 above)
Comment (shimpidipali) [reply celery_with_django/run_celery.bat:4]: I tried this but doesn't work.

**Comment by shimpidipali** [comment]
Added all changes

File: celery_with_django/django_celery_project/settings.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+# Application definition
+
+INSTALLED_APPS = [
+    'django.contrib.admin',
+    'django.contrib.auth',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
+    'django.contrib.staticfiles',
+
```
Comment (shimpidipali) [inline celery_with_django/django_celery_project/settings.py]: resolved

File: README.md [hunk 8]
```diff
@@ -1,2 +1,173 @@
...
+    source venv/bin/activate   # For Linux/Mac
+    .\venv\Scripts\activate    # For Windows
+    ```
+
+3. Install project dependencies:
+
+    ```bash
+    pip install -r requirements.txt
+    ```
+
```
Comment (shimpidipali) [inline README.md:65]: resolved

File: celery_with_django/main_app/tests.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (shimpidipali) [inline celery_with_django/main_app/tests.py]: resolved

File: celery_with_django/send_mail_app/admin.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (shimpidipali) [inline celery_with_django/send_mail_app/admin.py]: resolved

File: celery_with_django/main_app/tests.py (hunk 9 above)
Comment (shimpidipali) [inline celery_with_django/main_app/tests.py]: resolved

File: celery_with_django/django_celery_project/settings.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+CELERY_ACCEPT_CONTENT = ['application/json']
+CELERY_RESULT_SERIALIZER = os.getenv("CELERY_RESULT_SERIALIZER")
+CELERY_TASK_SERIALIZER = os.getenv("CELERY_TASK_SERIALIZER")
+CELERY_TIMEZONE = os.getenv("CELERY_TIMEZONE")
+CELERY_RESULT_EXTENDED =os.getenv("CELERY_RESULT_EXTENDED")
+CELERY_RESULT_BACKEND = os.getenv("CELERY_RESULT_BACKEND")
+CELERY_TASK_TRACK_STARTED = os.getenv("CELERY_TASK_TRACK_STARTED")
+
+
+# diffrent  queue according to task
```
Comment (shimpidipali) [inline celery_with_django/django_celery_project/settings.py]: resolved

File: README.md [hunk 12]
```diff
@@ -1,2 +1,173 @@
-# python_poc
-All the proof of concepts relevant to python tech
+# Django Project with Celery and Flower - README
+
+This is a Django project integrated with Celery for asynchronous task processing and Flower for real-time monitoring.
+
+## Prerequisites
+
+Before running this project, ensure you have the following prerequisites installed:
+
```
Comment (shimpidipali) [inline README.md:8]: resolved

File: celery_with_django/send_mail_app/tests.py [hunk 13]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (shimpidipali) [inline celery_with_django/send_mail_app/tests.py]: resolved

File: README.md [hunk 14]
```diff
@@ -1,2 +1,173 @@
...
+- -A django_celery_project.celery:
+   - Specifies the Celery application to use. -A stands for "app" or "application".
+   - In this case, django_celery_project.celery refers to the Celery application instance created in the celery.py module within the django_celery_project Django project.
+- worker:
+   - Indicates that this command will start a Celery worker process.
+   - Celery workers are responsible for executing tasks that are sent to them via the message broker.
+- -n worker_for_add_sub:
+   - Specifies the name of the Celery worker.
+    - Workers are identified by unique names, which can be  - - useful for monitoring and managing workers.
+    - In this case, the worker is named worker_for_add_sub.
```
Comment (shimpidipali) [inline README.md:157]: resolved

File: README.md [hunk 15]
```diff
@@ -1,2 +1,173 @@
...
+   - This is the command-line interface (CLI) for Celery, used to interact with Celery-related tasks and workers.
+- -A django_celery_project.celery:
+   - Specifies the Celery application to use. -A stands for "app" or "application".
+   - In this case, django_celery_project.celery refers to the Celery application instance created in the celery.py module within the django_celery_project Django project.
+- worker:
+   - Indicates that this command will start a Celery worker process.
+   - Celery workers are responsible for executing tasks that are sent to them via the message broker.
+- -n worker_for_add_sub:
+   - Specifies the name of the Celery worker.
+    - Workers are identified by unique names, which can be  - - useful for monitoring and managing workers.
```
Comment (shimpidipali) [inline README.md:156]: resolved

File: celery_with_django/django_celery_project/celery.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
+from __future__ import absolute_import, unicode_literals
+import os
+from celery import Celery
+from django.conf import settings
+from celery.schedules import crontab
+from time import sleep
```
Comment (shimpidipali) [inline celery_with_django/django_celery_project/celery.py]: resolved

File: celery_with_django/send_mail_app/views.py [hunk 17]
```diff
@@ -0,0 +1,3 @@
+from django.shortcuts import render
```
Comment (shimpidipali) [inline celery_with_django/send_mail_app/views.py]: resolved

File: README.md [hunk 18]
```diff
@@ -1,2 +1,173 @@
...
+
+    ```bash
+    python manage.py runserver
+    ```
+
+9. Access the Django admin interface and other endpoints in your browser.
+
+
+
+**Flow of celery and redis in django**
```
Comment (shimpidipali) [inline README.md:97]: resolved

File: README.md [hunk 19]
```diff
@@ -1,2 +1,173 @@
...
+# Django Project with Celery and Flower - README
+
+This is a Django project integrated with Celery for asynchronous task processing and Flower for real-time monitoring.
+
+## Prerequisites
+
+Before running this project, ensure you have the following prerequisites installed:
+
+1. **Python**: Make sure Python is installed on your system. You can download and install Python from [python.org](https://www.python.org/downloads/).
+
```
Comment (shimpidipali) [inline README.md:10]: resolved

File: celery_with_django/send_mail_app/admin.py (hunk 10 above)
Comment (shimpidipali) [inline celery_with_django/send_mail_app/admin.py]: resolved

File: celery_with_django/main_app/admin.py [hunk 20]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (shimpidipali) [inline celery_with_django/main_app/admin.py]: resolved

File: celery_with_django/send_mail_app/models.py [hunk 21]
```diff
@@ -0,0 +1,3 @@
+from django.db import models
```
Comment (shimpidipali) [inline celery_with_django/send_mail_app/models.py]: resolved

File: celery_with_django/django_celery_project/settings.py [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+
+# Build paths inside the project like this: BASE_DIR / 'subdir'.
+BASE_DIR = Path(__file__).resolve().parent.parent
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/5.0/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'django-insecure-w63dld2gxku7hkplgs%2i4hj3f0t_t8vrxxx0x6$u=pcq=g#7u'
```
Comment (shimpidipali) [inline celery_with_django/django_celery_project/settings.py]: resolved

File: README.md [hunk 23]
```diff
@@ -1,2 +1,173 @@
...
+    ```bash
+    pip install virtualenv
+    ```
+
+7. **Other Python Packages**: Depending on your project requirements, you may need to install additional Python packages. You can install them using pip.
+
+8. **Database Setup**: Configure your Django project to use a database. Django supports multiple databases like SQLite, PostgreSQL, MySQL, etc.
+
+9. **SMTP Server Configuration (Optional)**: If your project involves sending emails, configure the SMTP server settings in your Django project settings.
```
Comment (shimpidipali) [inline README.md:41]: resolved

File: celery_with_django/main_app/views.py [hunk 24]
```diff
@@ -0,0 +1,17 @@
+from django.http import HttpResponse
+from .tasks import add, sub
+from send_mail_app.tasks import send_mail_func
+
+def test(request):
+    # In Celery, the .delay() method is used to enqueue a task for asynchronous execution.
```
Comment (shimpidipali) [inline celery_with_django/main_app/views.py]: resolved

File: celery_with_django/main_app/views.py [hunk 25]
```diff
@@ -0,0 +1,17 @@
+from django.http import HttpResponse
+from .tasks import add, sub
+from send_mail_app.tasks import send_mail_func
+
+def test(request):
+    # In Celery, the .delay() method is used to enqueue a task for asynchronous execution. 
+    # When you define a Celery task using the @shared_task decorator,
```
Comment (shimpidipali) [inline celery_with_django/main_app/views.py]: resolved

File: celery_with_django/main_app/models.py [hunk 26]
```diff
@@ -0,0 +1,3 @@
+from django.db import models
```
Comment (shimpidipali) [inline celery_with_django/main_app/models.py]: resolved

File: celery_with_django/django_celery_project/celery.py [hunk 27]
```diff
@@ -0,0 +1,22 @@
+from __future__ import absolute_import, unicode_literals
+import os
+from celery import Celery
+from django.conf import settings
+from celery.schedules import crontab
```
Comment (shimpidipali) [inline celery_with_django/django_celery_project/celery.py]: resolved

File: README.md [hunk 28]
```diff
@@ -1,2 +1,173 @@
...
+    ```bash
+    pip install django
+    ```
+
+3. **Celery**: Install Celery using pip:
+
+    ```bash
+    pip install celery
+    ```
+
```
Comment (shimpidipali) [inline README.md:22]: resolved

File: celery_with_django/django_celery_project/celery.py (hunk 27 above)
Comment (shimpidipali) [inline celery_with_django/django_celery_project/celery.py]: resolved

File: celery_with_django/run_celery.bat (hunk 6 above)
Comment (reverseBitsTech) [reply celery_with_django/run_celery.bat:4]: Open a github issue regarding this, I am resolving the conversation

**Comment by reverseBitsTech** [comment]
Adding the PR changes to dev

**Pull request merged**